### ISSUE TITLE
Automatic download and decryption

### ISSUE THREAD

**Issue by mmiszczyk** [nlscc/samloader#24]
## Description
Usually, when using samloader, you want to download the latest software and decrypt it so that it can be flashed (or patched and then flashed). Now it requires three steps: checkupdate, download, decrypt. We could simplify it so that there is one command for doing this. We could add additional command which would:
- perform update check to get the most recent version
- download latest version, creating a file with the same name as the one present on the server
- decrypt it, with enc-version taken from file name (e.g. 2 for .enc2, 4 for .enc4 etc.)
- delete the encrypted archive if decryption is successful, leave it if not and output an error so that the user can deal with it himself

## Background
Doing the three-step check->download->decrypt seems to be a common workflow, at least it is for me. Having one command for that would be more convenient, and it wouldn't mess with other people's workflow as the other commands would not be changed.

**Comment by ianmacd** [comment]
I added a `mkfw` command to do just this in my fork of `samloader`. Unfortunately, the author rejected my pull request (in an earlier form).

I also have a `latest` command, which downloads and decrypts the latest version for any given device and region combination.

https://github.com/ianmacd/samloader/tree/devel

**Comment by name01019** [comment]
I agree with this. This needs to be much more streamlined
- The decryption method (2 or 4) should be automated from whether the filename has enc2 or enc4
- The decryption output file name should be autogenerated from the downloaded file
- Decryption should not require a region name. The downloaded file name should have a region name that can be passed on to the decryption command

**Comment by lekron42** [comment]
I created a small shell script which you can use directly on your android device using Termux which automates everything you mentioned.
For people who have to patch the firmware with Magisk, this is a really simple solution.

You can check it out here:
https://github.com/lekron42/samloader_termux

**Comment by 2ndMessiah** [comment]
> I added a `mkfw` command to do just this in my fork of `samloader`. Unfortunately, the author rejected my pull request (in an earlier form).
> 
> I also have a `latest` command, which downloads and decrypts the latest version for any given device and region combination.
> 
> https://github.com/ianmacd/samloader/tree/devel

yeah, such necessities r really easy to implement without disturbing the codes' stability...Dunno why the author rejects them all.

Might be sketching something big.

**Comment by hypernova7** [comment]
```bash
# only for linux
function samWizard () {
  local model=$1
  local region=$2
  local outputDir=${3:-$HOME/Downloads}
  checkupdate=$(samloader -m $model -r $region checkupdate)
  samloader -m $model -r $region download -DO $outputDir -v $checkupdate
  # -D argument is for decrypt after download, please check: samloader --help
}
samWizard "SM-A307G" "COL" $HOME/Downloads
```

**Comment by TheAirBlow** [comment]
[Syndical](https://github.com/TheAirBlow/Syndical) has that functionality built-in.
Feel free to test it!

**Comment by kassim** [comment]
Looks like this feature exists now with the flags `"-D", "--do-decrypt"`
 https://github.com/nlscc/samloader/blob/c25027f0e5f28b6b7c1db916fa296a8e72e0c729/samloader/main.py#L24

**Comment by nlscc** [comment]
this is part of the rewrite of main.py because the current interface and code quality is terrible

(@TheAirBlow please don't use this issue tracker to advertise your 'competing' project)

**Comment by TheAirBlow** [comment]
> this is part of the rewrite of main.py because the current interface and code quality is terrible
> 
> (@TheAirBlow please don't use this issue tracker to advertise your 'competing' project)

amongus
